ComponentTrait composing and trait CR naming design (#628)
oam-dev/kubevela: design/vela-core/componenttrait-composing-and-trait-CR-naming.md

# ComponentTrait Composing and Trait CR Naming

* Owner: Jianbo Sun (@wonderflow), Zhou Zheng Xi (@zzxwill)
* Reviewers: KubeVela/Crossplane Maintainers
* Status: Draft

## Background

> With the evolution of TraitDefinition, definition name can be short and simple without group and kind information,

Reviewer: ```suggestion
Now definition name is no longer coupled with CRD name, it's align to capability name  in KubeVela,
```

Author: Professional! Changing `it's align to` to `it's aligned with` might be better.

> and two TraitDefinitions can both refer to the same Trait CRD. So it's necessary to specify how to compose ComponentTrait

Author: 改成 `assemble trait CR from ApplicationConfiguration`? 这个地方还是讲怎么写 appconfig 里的 ComponentTrait 部分

> and two TraitDefinitions can both refer to the same Trait CRD. So it's necessary to specify how to compose ComponentTrait

Reviewer: 怎么写都可以，重点是写成让大家看懂的， compose ComponentTrait 以及 Trait CRD 是看不懂的部分，得改掉

in ApplicationConfiguration and to set the CR name in a friendly way.

> ## Several ways to compose ComponentTrait in ApplicationConfiguration

Reviewer: ```suggestion
## Several ways to assemble trait CR from ApplicationConfiguration
```

Author: from 是不是不对？

Reviewer: ## Several ways to assemble trait CR in ApplicationConfiguration

> - If the name of TraitDefinition is the same as that of Trait CRD

Reviewer: ```suggestion
- If the name of TraitDefinition is the same with the referenced CRD
```

后面还有一个trait变多个CR的情况，这里要撇开关系，没有Trait CRD 这种说法

> As shown in [OAM Kubernetes Runtime examples](https://github.com/crossplane/oam-kubernetes-runtime/blob/master/examples/containerized-workload/sample_application_config.yaml), the normal way to compose ComponentTrait is as below.

Reviewer: ```suggestion
The normal way to assemble a trait CR works as below.
```

```
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: example-component
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: ManualScalerTrait
            spec:
              replicaCount: 3
```

> - If the name of TraitDefinition is different to that of Trait CRD

Reviewer: 同上

The definition name `autoscale` is different to the trait name `autoscalers.standard.oam.dev`, we have two ways to compose ComponentTrait.

```
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: autoscale
spec:
  definitionRef:
    name: autoscalers.standard.oam.dev
```

1) Use label `trait.oam.dev/type`

```
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components: example-component
  traits:
    - trait:
        apiVersion: standard.oam.dev/v1alpha2
        kind: Autoscalers
        metadata:
          labels:

>             - oam.TraitTypeLabel: "autoscale"

Reviewer: ```suggestion
            trait.oam.dev/type: autoscale
```

代码和yaml混合谁看得懂？

        spec:
          replicaCount: 3
```

When rendering trait, the TraitDefinition could be retrieved by the label.

2) Use `name` field

```
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components: example-component
  traits:
    - trait:
        apiVersion: standard.oam.dev/v1alpha2
        kind: Autoscalers

>         name: autoscale

Reviewer: We never have this format. 

refer to https://github.com/oam-dev/spec/blob/master/7.application_configuration.md#trait

`name+properties` or `apiVersion+kind`

No mixing case

        spec:
          replicaCount: 3
```

The mutating handler will, at the very beginning,  convert `name:autoscale` to the labels above.

In summary, among these two ways to compose ComponentTrait, using labeling is recommended.

While in KubeVela, you don't need to worry about either `label` or `name`, just use command `vela TraitType` or `vela up` to attach the trait to
a component.

## Set the CR name in a friendly way

> Before OAM Kubernetes Runtime release v0.3.2, we named all trait name in the format of `${ComponentName}-trait-${HashTag}`. This will lead to confusions

Reviewer: ```suggestion
Currently, we named all trait name in the format of `${ComponentName}-trait-${HashTag}`. This will lead to confusions
```

> when listing CRs of all Traits of a component. So we propose to change the name to `${ComponentName}-${TraitDefinitionName}-${HashTag}`.

Reviewer: ```suggestion
when listing CRs of all Traits of a component. 

In this proposal, we propose to change the naming rule to `${ComponentName}-${TraitDefinitionName}-${HashTag}`.
```

> If the name of a TraitDefinition is in short style, like `autoscale`, we name the trait CR to `example-component-autoscale-xyzfa32r`.

Reviewer: ```suggestion
For example, if the name of TraitDefinition is `autoscale`, we name the trait CR to `example-component-autoscale-xyzfa32r`.
```

```
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components: example-component
  traits:
    - trait:
        apiVersion: standard.oam.dev/v1alpha2
        kind: Autoscalers
        metadata:
          labels:

>             - oam.TraitTypeLabel: "autoscale"

Reviewer: 同上

        spec:
          replicaCount: 3
```

If the name is the same as the CRD name, like `manualscalertraits.core.oam.dev`, we name it to `example-component-manualscalertraits-rq234rrw`

> by choosing the first part.

Reviewer: please give a specify example for this case as the before one
